### ISSUE TITLE
DataGrid is slow when cells contain a lot of (long) text

### ISSUE THREAD

**Issue by krassowski** [jupyterlab/lumino#448]
## Description

Rendering, and scrolling within DataGrid is very lagy when large text is present in cells. This is due to `CanvasRenderingContext2D.measureText` invocations.

- Firefox: 8.5 seconds
- Chrome: 11.5 seconds

Firefox:

![Screenshot from 2022-10-27 00-15-55](https://user-images.githubusercontent.com/5832902/198156234-df8e5c3c-ebcd-4f69-b570-e9c573e0a5bf.png)

![Screenshot from 2022-10-27 00-16-16](https://user-images.githubusercontent.com/5832902/198156218-885fd8e7-c89c-4785-9101-c0e08665457d.png)

## Reproduce

https://github.com/jupyterlab/jupyterlab-plugin-playground with:

<details>

```ts
import {
  JupyterFrontEnd,
  JupyterFrontEndPlugin,
} from '@jupyterlab/application';
import { MainAreaWidget } from '@jupyterlab/apputils';
import { DataGrid, DataModel, BasicKeyHandler, BasicMouseHandler, BasicSelectionModel } from '@lumino/datagrid';

class LargeDataModel extends DataModel {
  rowCount(region: DataModel.RowRegion): number {
    return region === 'body' ? 1000 : 1;
  }

  columnCount(region: DataModel.ColumnRegion): number {
    return region === 'body' ? 10 : 1;
  }

  data(region: DataModel.CellRegion, row: number, column: number): any {
    if (region !== 'body') {
      return `${row}, ${column}`;
    }
    return 'x'.repeat(10*1024);
  }
}

const extension: JupyterFrontEndPlugin<void> = {
  id: 'datagrid',
  autoStart: true,
  activate: (
    app: JupyterFrontEnd
  ) => {
    const model = new LargeDataModel();
    const content = new DataGrid();
  
    content.dataModel = model;
    // (handlers are not required for reproduction)
    content.keyHandler = new BasicKeyHandler();
    content.mouseHandler = new BasicMouseHandler();
    content.selectionModel = new BasicSelectionModel({
      dataModel: model
    });
    const widget = new MainAreaWidget({content});
    widget.id = 'datagrid-example';
    widget.title.label = 'Slow Datagrid Example';
    app.shell.add(widget, 'main');
  },
};

export default extension;
```

</details>

This is a bit pathological, but I got worse with just one column of long strings (CSS content of lab one sheet per row)

## Expected behavior

DataGrid loads fast (<2s for dataset of this size) and scrolls smoothly. For scroll, the text size is cached or something.

## Context

Only tested with lumino 1.x.

**Comment by fcollonval** [comment]
friendly ping to @martinRenou and @ibdafna

**Comment by martinRenou** [comment]
In the case of text wrapping enabled, we can perhaps come up with a better algorithm that uses less `measureText` calls.
We could also have some tricks to work around `measureText` slowness in the case of mono fonts.

**Comment by ibdafna** [comment]
Thanks for reporting! I'd argue this is an extreme case when each string has a length of 10240 characters - not justifying the performance, but just putting it out there. One simple fix would be to put a cap on the maximum string length we apply wrapping to and just disable the functionality if it's longer than the max.

**Comment by krassowski** [comment]
> I see there is one measureText call that we could probably get rid of if you don't need text wrapping.

Indeed, this one:

https://github.com/jupyterlab/lumino/blob/129221f9959d0f92e0c69159b464bd23cbc1f0a3/packages/datagrid/src/textrenderer.ts#L301-L324

redefines previous call from here:

https://github.com/jupyterlab/lumino/blob/15fb5d14136b9d465d744c3373bd347fe1f40bf0/packages/datagrid/src/textrenderer.ts#L217

So we could halve it easily.

> EDIT: I don't think there's text wrapping enabled in this case - just eliding (which also measures the text).

I agree, this looks like an issue in eliding, but when I skimmed in the code I saw that wrapping is implemented with naive linear search but could use binary search to reduce the cost to log(n):

https://github.com/jupyterlab/lumino/blob/129221f9959d0f92e0c69159b464bd23cbc1f0a3/packages/datagrid/src/textrenderer.ts#L238-L291

**Comment by ibdafna** [comment]
We should look to improve the wrapping algorithm for sure, but I guess my point here is that we should be eliminating or significantly speed up the `measureText` call which happens regardless or wrapping/eliding

**Comment by krassowski** [comment]
Not sure how much we can speed up `measureText` as it is pretty close to the metal.

Eliding a sentence with 1024 characters to a 1-character size column with current binary search takes 10  `measureText` calls. We could reduce it to 1 call (on the happy path) with:

```ts
// If dealing with a very large string
if (textWidth > 4 * boxWidth) {
   const approximateLengthToKeep = Math.ceil(text.length * boxWidth/textWidth);
   const candidateText = text.substring(0, approximateLengthToKeep + 1) + elide;
   const candidateWidth = gc.measureText().width;
   // Keep the result if we did not overshoot because of unfavourable string composition;
   // the old algorithm will fine-tune it to remove the few remaining extra characters
   if (newWidth >= boxWidth) {
      text = candidateText;
      textWidth = newWidth;
   } else {
      // no-op
      // we would end up here for strings like: `iiiiimmmmmm` if we are unlucky
   }
}
// keep old algorithm below as it was
while (textWidth > boxWidth && text.length > 1) { 
     if (text.length > 4 && textWidth >= 2 * boxWidth) { 
       // If text width is substantially bigger, take half the string 
       text = text.substring(0, text.length / 2 + 1) + elide; 
     } else { 
       // Otherwise incrementally remove the last character 
       text = text.substring(0, text.length - 2) + elide; 
     } 
     textWidth = gc.measureText(text).width; 
   } 
 }
```

The bigger issue is a wide column with the text, say equivalent to 200 characters (this is what I saw locally). Then if you start with 399 characters you would call `measureText` 199 times. This looks like a realistic scenario for many users: you add a column with one paragraph but want to display only the size of about one sentence.

**Comment by krassowski** [comment]
> The bigger issue is a wide column with the text, say equivalent to 200 characters (this is what I saw locally). Then if you start with 399 characters you would call measureText 199 times. This looks like a realistic scenario for many users: you add a column with one paragraph but want to display only the size of about one sentence.

Example in details:

<details>

```ts
import {  JupyterFrontEnd, JupyterFrontEndPlugin } from '@jupyterlab/application';
import { MainAreaWidget } from '@jupyterlab/apputils';
import { DataGrid, DataModel, BasicKeyHandler, BasicMouseHandler, BasicSelectionModel } from '@lumino/datagrid';

class LargeDataModel extends DataModel {
  rowCount(region: DataModel.RowRegion): number {
    return region === 'body' ? 100 : 1;
  }

  columnCount(region: DataModel.ColumnRegion): number {
    return region === 'body' ? 1 : 1;
  }

  data(region: DataModel.CellRegion, row: number, column: number): any {
    if (region !== 'body') {
      return `${row}, ${column}`;
    }
    return 'x'.repeat(250);
  }
}

const extension: JupyterFrontEndPlugin<void> = {
  id: 'datagrid',
  autoStart: true,
  activate: (
    app: JupyterFrontEnd
  ) => {
    const model = new LargeDataModel();
    const content = new DataGrid();
  
    content.dataModel = model;
    // (handlers are not required for reproduction)
    content.keyHandler = new BasicKeyHandler();
    content.mouseHandler = new BasicMouseHandler();
    content.selectionModel = new BasicSelectionModel({
      dataModel: model
    });
    content.resizeColumn('body', 0, 800);
    const widget = new MainAreaWidget({content});
    widget.id = 'datagrid-example';
    widget.title.label = 'Slow Datagrid Example';
    app.shell.add(widget, 'main');
  },
};

export default extension;
```

![Screenshot from 2022-10-27 20-38-37](https://user-images.githubusercontent.com/5832902/198383268-fd6618f3-7967-40f2-9cd5-2e30da00fc66.png)

Loading freezes the UI for 1.5 seconds on Chrome.

</details>

**Comment by krassowski** [comment]
Yes I agree that the issue I saw is not in `wrapText` (I never said it was).

As for caching, it might help a bit with scrolling (at cost of doubling the RAM if not much worse) but as soon as user resizes the column we are back at square one.

Just to spell it out, I was referring to eliding in last few comments and eliding is enabled by default and inefficient in the worst case scenario (as described above).

**Comment by ibdafna** [comment]
Okay, understood 😊 I think I have some ideas about how to fix this. Shouldn't be anything difficult. Will report back!

**Comment by ibdafna** [comment]
@krassowski @fcollonval I've done more profiling to see what the performance is like without eliding or wrapping, and unfortunately this is still quite laggy even with zero calls to `measureText`. The `fillText` call seems to take a long time.

![lumino_perf](https://user-images.githubusercontent.com/24281433/199163206-d7f04a7d-ec5e-424d-a86a-dec913cc7248.gif)

**Comment by ibdafna** [comment]
Digging deeper, I think this may have to do with our rendering optimizations. When resizing column/row headers or scrolling, we trigger a full on-screen buffer paint - this is extremely slow. We should be writing to the off-screen buffer and copying any changes from that buffer to the on-screen one. 

![lumini_perf_2](https://user-images.githubusercontent.com/24281433/199164623-0d44d60f-82a6-4ed5-ad71-fd4957fa38c2.gif)
